### PR TITLE
periodically checking for new tracking statements

### PR DESCRIPTION
**Pull request by oconnor663**
I want to get comments from @maxtaco about whether this is a reasonable way to implement this feature:
1) we look for new tracking statements in every LoadMe call (so the check-tracks interface just calls LoadMe)
2) we store the last seen seqno in LevelDB

Also @chrisnojima mentioned that we want to check proof state in addition to new signatures. That could be expensive (because it's O(people you track) instead of O(1)), so I'll want to talk about that with @cjb and you guys.

### CODE REVIEW

File: go/engine/check_tracks.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
```
Comment (maxtaco) [inline go/engine/check_tracks.go]: I'd say nuke this since it's almost 100% boilerplate.

File: go/libkb/id_table.go [hunk 2]
```diff
@@ -389,25 +390,32 @@ func (l TrackChainLink) IsRemote() bool {
 }
 
 func ParseTrackChainLink(b GenericChainLink) (ret *TrackChainLink, err error) {
-	var whom string
-	whom, err = b.payloadJSON.AtPath("body.track.basics.username").GetString()
+	var whomUsername string
+	whomUsername, err = b.payloadJSON.AtPath("body.track.basics.username").GetString()
```
Comment (maxtaco) [inline go/libkb/id_table.go]: all `GetUID` on a `jsonw.Wrapper`

File: go/libkb/loaduser.go [hunk 3]
```diff
@@ -360,3 +362,64 @@ func LoadUserPlusKeys(g *GlobalContext, uid keybase1.UID, cacheOK bool) (keybase
...
+		if untrackChainLink, ok := link.(*UntrackChainLink); ok {
+			arg.G().Log.Debug("| found new untrack %s", untrackChainLink.whomUID)
+			changedUsers[untrackChainLink.whomUID] = true
+		}
+	}
+
+	// Fire off user changed notifications for anyone referred to by a new
+	// track or untrack link.
+	for uid := range changedUsers {
+		arg.G().NotifyRouter.HandleUserChanged(uid)
```
Comment (maxtaco) [inline go/libkb/loaduser.go]: We should use a  new notification channel for this

File: go/libkb/loaduser.go [hunk 4]
```diff
@@ -360,3 +362,64 @@ func LoadUserPlusKeys(g *GlobalContext, uid keybase1.UID, cacheOK bool) (keybase
...
+	arg.G().Log.Debug("+ checkForNewTracksIfSelf")
+	defer arg.G().Log.Debug("- checkForNewTracksIfSelf")
+
+	// If we're loading a user other than ourselves, short-circuit.
+	if !arg.Self || idtable == nil {
+		return nil
+	}
+
+	// Get the our last seen seqno from the local DB.
+	dbkey := DbKeyUID(DBLastSeenSeqno, arg.UID)
```
Comment (maxtaco) [inline go/libkb/loaduser.go]: Shouldn't this be available as a by-product as the loading process?  Do we need to keep track of this separately?

File: go/libkb/loaduser.go [hunk 5]
```diff
@@ -360,3 +362,64 @@ func LoadUserPlusKeys(g *GlobalContext, uid keybase1.UID, cacheOK bool) (keybase
...
+			return err
+		}
+		lastSeenSeqno = Seqno(lastSeenSeqnoInt)
+		arg.G().Log.Debug("| last seen seqno is %d", lastSeenSeqno)
+	}
+
+	// Find all the new track and untrack links since the last link we saw, and
+	// make a set of the users they refer to.
+	newSeenSeqno := lastSeenSeqno
+	changedUsers := make(map[keybase1.UID]bool)
```
Comment (maxtaco) [inline go/libkb/loaduser.go]: The more go-y way is, I've since learned `map[keybase1.UID]struct{}`

File: go/libkb/loaduser.go [hunk 6]
```diff
@@ -360,3 +362,64 @@ func LoadUserPlusKeys(g *GlobalContext, uid keybase1.UID, cacheOK bool) (keybase
...
+	for _, link := range idtable.links {
+		if link.GetSeqno() <= lastSeenSeqno {
+			continue
+		}
+		if link.GetSeqno() > newSeenSeqno {
+			newSeenSeqno = link.GetSeqno()
+		}
+		if trackChainLink, ok := link.(*TrackChainLink); ok {
```
Comment (maxtaco) [inline go/libkb/loaduser.go]: Can you use some sort of interface here, rather than type-casting for particular links?

**Comment by maxtaco** [comment]
I don't agree we need to check proof state, at least not at first.  Seems like this is a decent first-order approximation.

Are there any tests?

File: go/service/main.go [hunk 7]
```diff
@@ -184,6 +187,17 @@ func (d *Service) writeServiceInfo() error {
 	return rtInfo.WriteFile(path.Join(runtimeDir, "keybased.info"))
 }
 
+func (d *Service) checkTracksEveryHour() {
```
Comment (maxtaco) [inline go/service/main.go]: Is there a mechanism to rate-limit these checks if they happen more frequently than some lower limit?  You can imagine having 10 windows open all at once, each asking to check once a minute, which means you'll get 10 checks a minute in the simple implementation.

File: go/libkb/loaduser.go [hunk 8]
```diff
@@ -192,6 +192,8 @@ func LoadUser(arg LoadUserArg) (ret *User, err error) {
 		err = NoKeyError{emsg}
 	}
 
+	checkForNewTracksIfSelf(arg, ret.IDTable())
```
Comment (maxtaco) [inline go/libkb/loaduser.go]: the `LoadUser` call should short-circuit if it realizes that it has a fresh version (after the merkle leaf check). No reason to load the rest of the user just to throw it on the floor.

Similarly, use the freshness check in sigchain loading to deduce what the previously-save chain tail was, and what the current chain-tail is, and feel that data through to `checkForNewTrackIfSelf` to remove the need for redundant level-DB storage of this info.

File: go/client/cmd_check_tracks.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+type CmdCheckTracks struct {
+	libkb.Contextified
+}
+
+func (c *CmdCheckTracks) ParseArgv(ctx *cli.Context) error {
+	return nil
+}
+
+func (c *CmdCheckTracks) Run() (err error) {
+	cli, err := GetTrackClient()
```
Comment (maxtaco) [inline go/client/cmd_check_tracks.go]: this must be contextified.

File: go/service/main.go (hunk 7 above)
Comment (cjb) [inline go/service/main.go]: I've suggested rate-limiting too, but FWIW, on the Electron side I plan to send one RPC per minute when *any* tracker window is open, not one RPC per window per minute.

**Comment by oconnor663** [comment]
@maxtaco this is ready for another review. New code of interest: the `ratelimits.go` and `systests/tracking_test.go`.

File: go/engine/fakeidentifyui.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
```
Comment (oconnor663) [inline go/engine/fakeidentifyui.go]: Let's call this `test_fakeidentifyui.go`.

File: go/libkb/ratelimit.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+const (
+	CheckTrackingRateLimit RateLimitCategory = "CheckTrackingRateLimit"
+	TestEventRateLimit                       = "TestEventRateLimit"
+)
+
+type RateLimits struct {
+	lastActionTimes map[RateLimitCategory]time.Time
+	mutex           *sync.Mutex
+	Contextified
```
Comment (oconnor663) [inline go/libkb/ratelimit.go]: put this at the top

File: go/libkb/ratelimit.go [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+type RateLimitCategory string
+
+const (
+	CheckTrackingRateLimit RateLimitCategory = "CheckTrackingRateLimit"
+	TestEventRateLimit                       = "TestEventRateLimit"
+)
+
+type RateLimits struct {
+	lastActionTimes map[RateLimitCategory]time.Time
+	mutex           *sync.Mutex
```
Comment (oconnor663) [inline go/libkb/ratelimit.go]: get rid of the name, and don't forget to initialize

File: go/client/cmd_check_tracking.go [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+	err = cli.CheckTracking(context.TODO(), 0 /* session ID */)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func NewCmdCheckTracking(cl *libcmdline.CommandLine, g *libkb.GlobalContext) cli.Command {
+	return cli.Command{
+		Name:  "check-tracks",
```
Comment (oconnor663) [inline go/client/cmd_check_tracking.go]: change this to "check-tracking"

File: go/libkb/ratelimit.go [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+func NewRateLimits(g *GlobalContext) *RateLimits {
+	return &RateLimits{
+		lastActionTimes: make(map[RateLimitCategory]time.Time),
+		Contextified:    NewContextified(g),
+	}
+}
+
+func (r *RateLimits) GetPermission(category RateLimitCategory, interval time.Duration) bool {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
```
Comment (oconnor663) [inline go/libkb/ratelimit.go]: `r.Lock` etc.

File: go/systests/tracking_test.go [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+func (h *trackingNotifyHandler) TrackingChanged(_ context.Context, uid keybase1.UID) error {
+	h.trackingCh <- uid
+	return nil
+}
+
+func TestTrackingNotifications(t *testing.T) {
+	tc := setupTest(t, "signup")
+
+	// Hack the various portions of the service that aren't
+	// properly contextified.
```
Comment (oconnor663) [inline go/systests/tracking_test.go]: Add back in the `	libkb.G.LocalDb = nil` line. Also take a look at using the same Cmd* interfaces that the user_test.go does.

File: go/systests/tracking_test.go [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+	err := trackCmd.Run()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Do a check for new tracking statements that should fire off a
+	// notification. Currently the TrackEngine above does not fetch the new
+	// chain link from the server, so this call is required. It's possible that
+	// TrackEngine (or our signature caching code) might change in the future,
+	// making this call unnecessary.
```
Comment (oconnor663) [inline go/systests/tracking_test.go]: Update comment.

**Comment by oconnor663** [comment]
Landed as ed0d37f45322b481c52ea2f142969a59d935c2c4.

**Pull request closed without merging**